File: backend/atlas/spaces/content.py
"""Space content + persistence (Phase B, increment 2). A post is content authored by a MEMBER
(role-gated via the authority engine), committed by hash, and witnessed per the space's persistence
mode — composing the primitives that already exist, not new crypto.

Persistence modes (orthogonal; escalation Present → Fading → Private → Public):

  * PRESENT — live only; NO stored copy. The item exists in the moment; nothing is recorded.
  * FADING  — stored with an expiry epoch; pruned once `now` passes it.
  * PRIVATE — the commitment is appended to the space's IndividualLedger; provable BY THE PARTIES
              (inclusion proof against a root they share), not to the world.
  * PUBLIC  — additionally anchors the ledger root to the GlobalAnchorLog; provable to ANYONE.

Content confidentiality (sealing the bytes) is the existing `spaces/space.py` vault; this module is
the durability/witness layer over the commitment.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, List, Optional, Sequence

from ..crypto.primitives import H
from ..ledger.global_anchor import GlobalAnchorLog
from ..ledger.individual import IndividualLedger
from .kinds import Access, IdentityTier, PersistenceMode, Role, SpaceDescriptor, has_role


class AccessError(Exception):
    """Posting was refused — the author failed this space's access gate (fail-closed)."""


@dataclass(frozen=True)
class Ban:
    """A moderation record: `target` (a persona handle) is barred from an OPEN space. Honored only
    when issued by a MODERATOR+ (enforced in `SpaceStore.ban`). Block-list moderation — the inverse of
    an invite: OPEN spaces admit everyone verified, then remove abusers, rather than admitting per-person."""

    space_id: bytes
    target: bytes
    epoch: int


def content_commitment(space_id: bytes, author: bytes, content: bytes, parent: bytes = b"") -> bytes:
    """The public commitment to a post: binds the space, the author handle, the content, and (for a
    comment/reply) the `parent` item it hangs under. `parent=b""` for a top-level post."""
    return H(b"atlas/space-item", space_id, author, content, parent)


@dataclass(frozen=True)
class SpaceItem:
    space_id: bytes
    author: bytes                 # persona handle (opaque)
    content_hash: bytes
    persistence: PersistenceMode
    seq: int
    expiry: Optional[int] = None  # set only for FADING
    parent: Optional[bytes] = None  # content_hash of the item this replies to (None = top-level post)


class SpaceStore:
    """Content store for ONE space. Dispatches by persistence mode over the existing ledgers. Holds
    commitments + metadata only — the content bytes are sealed in the space vault elsewhere."""

    def __init__(self, space: SpaceDescriptor, *, global_anchor: Optional[GlobalAnchorLog] = None,
                 is_verified_human: Optional[Callable[[bytes], bool]] = None):
        self.space = space
        self.ledger = IndividualLedger(owner_id=space.space_id)   # PRIVATE / PUBLIC records
        self.global_anchor = global_anchor                        # shared; required for PUBLIC
        # OPEN spaces gate posting on VERIFIED-HUMANNESS (the sybil defense) instead of an invite. The
        # predicate is keyed by author handle; without it, OPEN posting is fail-closed (nobody passes).
        self.is_verified_human = is_verified_human
        self._banned: set = set()                                 # persona handles barred (block-list)
        self._items: List[SpaceItem] = []
        self._seq = 0

    def ban(self, mod_chain: Sequence, target: bytes, *, now: int, epoch: int = 0) -> Ban:
        """MODERATE an OPEN space by BLOCK-list: bar `target` from posting. Authorized only for a
        MODERATOR+ (fail-closed via the authority engine) — a random persona can't ban a rival."""
        if not has_role(self.space, mod_chain, at_least=Role.MODERATOR, now=now):
            raise AccessError("banning requires >= MODERATOR")
        self._banned.add(target)
        return Ban(space_id=self.space.space_id, target=target, epoch=epoch)

    def _gate(self, author_chain, author: bytes, now: int) -> None:
        """Fail-closed posting gate: TWO orthogonal checks — ACCESS (who may enter) then IDENTITY
        (what accountability the space demands)."""
        access = self.space.access
        # 1) ACCESS — the concentric rings.
        if access == Access.OPEN:
            if author in self._banned:                        # public square: block-list moderation
                raise AccessError("author is banned from this space")
        elif access == Access.SELF:
            if not has_role(self.space, author_chain, at_least=Role.OWNER, now=now):
                raise AccessError("SELF space: only the owner may post")
        else:                                                 # INVITE / MEMBER: allow-list
            if not has_role(self.space, author_chain, at_least=Role.MEMBER, now=now):
                raise AccessError("author needs >= MEMBER to post")

        # 2) IDENTITY — the accountability tier (orthogonal to access).
        identity = self.space.identity
        if identity == IdentityTier.VERIFIED_PERSON:
            # one-human-one-voice sybil gate — a personhood-backed pseudonym is required.
            if self.is_verified_human is None or not self.is_verified_human(author):
                raise AccessError("this space requires a verified-person identity")
        elif identity == IdentityTier.PSEUDONYMOUS:
            if not author:                                    # a persistent pseudonym (any handle)
                raise AccessError("this space requires a pseudonym")
        # IdentityTier.ANONYMOUS — no identity requirement.

    def post(self, author_chain, author: bytes, *, content: bytes, now: int,
             persistence: Optional[PersistenceMode] = None, ttl: Optional[int] = None,
             parent: Optional[bytes] = None) -> SpaceItem:
        """Author a post — or, when `parent` is the content_hash of another item, a COMMENT/reply that
        threads under it (a comment is just a post with a parent; same access gate, same persistence).
        The gate depends on the space's ACCESS tier (SELF / INVITE / MEMBER = allow-list via the
        authority engine; OPEN = verified-human + not-banned)."""
        self._gate(author_chain, author, now)
        mode = persistence if persistence is not None else self.space.persistence
        commit = content_commitment(self.space.space_id, author, content, parent or b"")
        self._seq += 1
        expiry = (now + ttl) if (mode == PersistenceMode.FADING and ttl is not None) else None
        item = SpaceItem(space_id=self.space.space_id, author=author, content_hash=commit,
                         persistence=mode, seq=self._seq, expiry=expiry, parent=parent)

        if mode == PersistenceMode.PRESENT:
            pass                                              # live only — nothing stored
        elif mode == PersistenceMode.FADING:
            self._items.append(item)                          # stored; pruned in `live()`
        else:                                                 # PRIVATE or PUBLIC
            self.ledger.append(commit)                        # ledgered between the parties
            self._items.append(item)
            if mode == PersistenceMode.PUBLIC:
                if self.global_anchor is None:
                    raise AccessError("PUBLIC persistence requires a global anchor")
                self.global_anchor.anchor(self.space.space_id, self.ledger.root,
                                          drand_round=now.to_bytes(8, "big"))
        return item

    def live(self, now: int) -> List[SpaceItem]:
        """Currently-live items — FADING items past their expiry are pruned."""
        return [it for it in self._items if it.expiry is None or now <= it.expiry]

    def replies(self, parent: bytes) -> List[SpaceItem]:
        """The comments/replies threaded directly under `parent` (its content_hash), in post order."""
        return [it for it in self._items if it.parent == parent]

    def is_publicly_provable(self) -> bool:
        """Was the current ledger root anchored to the global log (provable to anyone)?"""
        return (self.global_anchor is not None
                and self.global_anchor.is_anchored(self.space.space_id, self.ledger.root))

File: ios/AtlasCore/Sources/AtlasCore/Spaces/Content.swift
import Foundation

/// Space content + persistence — mirrors `backend/atlas/spaces/content.py`. A post is content by a
/// role-gated author, committed by hash, and witnessed per the space's persistence mode
/// (PRESENT → FADING → PRIVATE → PUBLIC). Composes existing primitives — the authority engine
/// (`hasRole`), `IndividualLedger`, `GlobalAnchor.Log` — no new crypto. Holds commitments +
/// metadata only; content bytes are sealed in the space vault elsewhere.
extension Spaces {

    public enum ContentError: Error, Equatable { case access(String) }

    /// The public commitment to a post: binds space, author handle, content, and (for a reply) the
    /// `parent` it threads under. `parent` empty (default) for a top-level post.
    public static func contentCommitment(spaceID: Data, author: Data, content: Data,
                                         parent: Data = Data()) -> Data {
        Primitives.H(Data("atlas/space-item".utf8), spaceID, author, content, parent)
    }

    /// A moderation record: `target` (a persona handle) barred from an OPEN space. Honored only when
    /// issued by a MODERATOR+ (enforced in `SpaceStore.ban`) — block-list moderation.
    public struct Ban: Equatable {
        public let spaceID: Data
        public let target: Data
        public let epoch: UInt64
    }

    public struct SpaceItem: Equatable {
        public let spaceID: Data
        public let author: Data              // persona handle (opaque)
        public let contentHash: Data
        public let persistence: PersistenceMode
        public let seq: Int
        public let expiry: UInt64?           // set only for FADING
        public let parent: Data?             // content hash of the item this replies to (nil = top-level)
    }

    /// Content store for ONE space. Dispatches by persistence mode over the existing ledgers.
    public final class SpaceStore {
        public let space: SpaceDescriptor
        private let ledger: IndividualLedger
        private let globalAnchor: GlobalAnchor.Log?
        private let isVerifiedHuman: ((Data) -> Bool)?
        private var banned: Set<Data> = []
        private var items: [SpaceItem] = []
        private var seq = 0

        public init(space: SpaceDescriptor, globalAnchor: GlobalAnchor.Log? = nil,
                    isVerifiedHuman: ((Data) -> Bool)? = nil) {
            self.space = space
            self.ledger = IndividualLedger(ownerID: space.spaceID)
            self.globalAnchor = globalAnchor
            self.isVerifiedHuman = isVerifiedHuman
        }

        /// MODERATE an OPEN space by block-list: bar `target`. Authorized only for MODERATOR+
        /// (fail-closed via the authority engine) — a random persona can't ban a rival.
        @discardableResult
        public func ban(modChain: [Authority.Grant], target: Data, now: UInt64, epoch: UInt64 = 0) throws -> Ban {
            guard Spaces.hasRole(space, modChain, atLeast: .moderator, now: now) else {
                throw ContentError.access("banning requires >= MODERATOR")
            }
            banned.insert(target)
            return Ban(spaceID: space.spaceID, target: target, epoch: epoch)
        }

        /// Fail-closed posting gate: ACCESS (who may enter) then IDENTITY (accountability tier).
        private func gate(_ authorChain: [Authority.Grant], _ author: Data, _ now: UInt64) throws {
            switch space.access {
            case .open:
                if banned.contains(author) { throw ContentError.access("author is banned from this space") }
            case .selfOnly:
                if !Spaces.hasRole(space, authorChain, atLeast: .owner, now: now) {
                    throw ContentError.access("SELF space: only the owner may post")
                }
            default: // .invite / .member — allow-list
                if !Spaces.hasRole(space, authorChain, atLeast: .member, now: now) {
                    throw ContentError.access("author needs >= MEMBER to post")
                }
            }
            switch space.identity {
            case .verifiedPerson:
                if isVerifiedHuman == nil || !(isVerifiedHuman!(author)) {
                    throw ContentError.access("this space requires a verified-person identity")
                }
            case .pseudonymous:
                if author.isEmpty { throw ContentError.access("this space requires a pseudonym") }
            case .anonymous:
                break
            }
        }

        /// Author a post — or, when `parent` is another item's content hash, a COMMENT/reply that
        /// threads under it (same gate, same persistence).
        @discardableResult
        public func post(authorChain: [Authority.Grant], author: Data, content: Data, now: UInt64,
                         persistence: PersistenceMode? = nil, ttl: UInt64? = nil,
                         parent: Data? = nil) throws -> SpaceItem {
            try gate(authorChain, author, now)
            let mode = persistence ?? space.persistence
            let commit = Spaces.contentCommitment(spaceID: space.spaceID, author: author,
                                                  content: content, parent: parent ?? Data())
            seq += 1
            let expiry: UInt64? = (mode == .fading && ttl != nil) ? now + ttl! : nil
            let item = SpaceItem(spaceID: space.spaceID, author: author, contentHash: commit,
                                 persistence: mode, seq: seq, expiry: expiry, parent: parent)
            switch mode {
            case .present:
                break                                       // live only — nothing stored
            case .fading:
                items.append(item)                          // stored; pruned in live()
            case .privateMode, .publicMode:
                ledger.append(commit)                       // ledgered between the parties
                items.append(item)
                if mode == .publicMode {
                    guard let anchor = globalAnchor else {
                        throw ContentError.access("PUBLIC persistence requires a global anchor")
                    }
                    var be = now.bigEndian
                    let round = withUnsafeBytes(of: &be) { Data($0) }
                    _ = try anchor.anchor(ownerID: space.spaceID, root: ledger.root, epochRound: round)
                }
            }
            return item
        }

        /// Currently-live items — FADING items past their expiry are pruned.
        public func live(now: UInt64) -> [SpaceItem] {
            items.filter { $0.expiry == nil || now <= $0.expiry! }
        }

        /// The comments/replies threaded directly under `parent` (its content hash), in post order.
        public func replies(parent: Data) -> [SpaceItem] {
            items.filter { $0.parent == parent }
        }

        /// Was the current ledger root anchored to the global log (provable to anyone)?
        public func isPubliclyProvable() -> Bool {
            guard let anchor = globalAnchor else { return false }
            return anchor.isAnchored(ownerID: space.spaceID, root: ledger.root)
        }
    }
}
